File: web/js/lightdm.js
'use strict';

const DEBUG_PASSWORD = 'test';

window.debug = window.lightdm === undefined;

if (window.debug)
{
    window.lightdm = {
        is_authenticated: false,
        authentication_user: undefined,
        default_session: 'plasma-shell',
        sessions: [
            {
                name: 'KDE 5',
                key: 'plasma-shell'
            },
            {
                name: 'Gnome 3',
                key: 'gnome-shell'
            },
            {
                name: 'XFCE 4',
                key: 'xfce'
            }
        ],
        users: [{
            display_name: 'John Doe',
            username: 'johnd'
        }, {
            display_name: 'Adrien Navratil',
            username: 'litarvan',
            image: 'images/litarvan.png'
        }],
        languages: [{
            name: 'American English',
            code: 'en_US.utf8'
        }, {
            name: 'Français',
            code: 'fr_FR.utf8'
        }],
        language: 'American English',
        start_authentication: (username) => {
            console.log(`Starting authenticating : '${username}'`);
            lightdm.authentication_user = username;

            show_prompt("Password: ");
        },
        cancel_authentication: () => {
            console.log('Auth cancelled');
        },
        respond: (password) => {
            console.log(`Password provided : '${password}'`);

            if (password === DEBUG_PASSWORD)
            {
                lightdm.is_authenticated = true;
            }
            else
            {
                let now = new Date().getTime();
                while (new Date().getTime() < now + 2000);
            }

            authentication_complete();
        },
        login: (user, session) => {
            alert(`Logged with '${user}' (Session: '${session}') !`);
        },
        shutdown: () => {
            alert('(DEBUG: System is shutting down)')
        },
        restart: () => {
            alert('(DEBUG: System is rebooting)')
        }
    };
}

let password;
let errorCB;
let completeCB;

function login(username, pass, cb, errCB)
{
    completeCB = cb;
    errorCB = errCB;
    password = pass;

    lightdm.start_authentication(username);
}

function start(desktop)
{
    lightdm.login(lightdm.authentication_user, desktop);
}

function show_prompt(text, type)
{
    if (text === "Password: ")
    {
        lightdm.respond(password);
    }
}

function authentication_complete()
{
    if (lightdm.is_authenticated)
    {
        completeCB();
    }
    else
    {
        lightdm.cancel_authentication();
        errorCB('Invalid username/password');
    }
}

function show_message(text, type)
{
    errorCB(text);
}

File: packages/lightdm_webkit_theme_litarvan/lightdm.dart
import 'dart:js';

import 'package:angular2/core.dart';
import 'package:js/js.dart';
import 'package:lightdm_webkit_theme_litarvan/desktop.dart';
import 'package:lightdm_webkit_theme_litarvan/user.dart';

@Injectable()
class LightDM {
  JsObject lightdm;

  List<Desktop> _desktops;
  List<User> _users;
  List<String> _languages;
  String _language;

  LightDM() {
    this.lightdm = context['lightdm'];

    _desktops = [];
    _users = [];
    _languages = [];

    for (JsObject obj in lightdm['sessions']) {
      _desktops.add(new Desktop(obj['key'], obj['name']));
    }

    for (JsObject obj in lightdm['users']) {
      _users.add(new User(obj['display_name'], obj['username'], obj['image'] != null ? (obj['image'].startsWith('/') ? 'file://' : '') + obj['image'] : 'images/default_user.png'));
    }

    for (JsObject obj in lightdm['languages']) {
      _languages.add(obj['code']);
    }

    _language = lightdm['language'];
    print(_language);
  }

  login(String username, String password, Function completeCB, Function errCB) => context.callMethod('login', [username, password, allowInterop(completeCB), allowInterop(errCB)]);
  start(String desktop) => context.callMethod('start', [desktop]);
  shutdown() => lightdm.callMethod('shutdown');
  reboot() => lightdm.callMethod('restart');

  List<Desktop> get desktops => _desktops;
  List<User> get users => _users;
  List<String> get languages => _languages;
  String get language => _language;
}
